sdk/src/Services/S3/Custom/Transfer/Internal/BufferedPartDataHandler.cs

```diff
@@ -0,0 +1,141 @@
+    {
+        private readonly IPartBufferManager _partBufferManager;
+        private readonly BufferedDownloadConfiguration _config;
+        
+        public BufferedPartDataHandler(
+            IPartBufferManager partBufferManager,
+            BufferedDownloadConfiguration config)
+        {
+            _partBufferManager = partBufferManager ?? throw new ArgumentNullException(nameof(partBufferManager));
+            _config = config ?? throw new ArgumentNullException(nameof(config));
+        }
+        
+        public async Task ProcessPartAsync(
+            int partNumber,
+            GetObjectResponse response, 
+            CancellationToken cancellationToken)
+        {
+            // Buffer the part from the response stream into memory
+            var buffer = await BufferPartFromResponseAsync(
+                partNumber, 
+                response, 
+                cancellationToken).ConfigureAwait(false);
+                
+            // Add the buffered part to the buffer manager
+            await _partBufferManager.AddBufferAsync(buffer, cancellationToken).ConfigureAwait(false);
+        }
+        
+        public Task WaitForCapacityAsync(CancellationToken cancellationToken)
+        {
+            return _partBufferManager.WaitForBufferSpaceAsync(cancellationToken);
+        }
+        
+        public void ReleaseCapacity()
+        {
+            _partBufferManager.ReleaseBufferSpace();
+        }
+        
+        public void OnDownloadComplete(Exception exception)
+        {
+            _partBufferManager.MarkDownloadComplete(exception);
+        }
+        
+        public void Dispose()
+        {
+            // _partBufferManager is owned by caller, don't dispose
+        }
+        
+        private async Task<StreamPartBuffer> BufferPartFromResponseAsync(
+            int partNumber, 
+            GetObjectResponse response, 
+            CancellationToken cancellationToken)
+        {
+            StreamPartBuffer downloadedPart = null;
+            
+            try
+            {
+                // Use ContentLength to determine exact bytes to read and allocate
+                long expectedBytes = response.ContentLength;
+                int initialBufferSize = (int)expectedBytes;
+                
+                downloadedPart = StreamPartBuffer.Create(partNumber, initialBufferSize);
+                
+                // Get reference to the buffer for writing
+                var partBuffer = downloadedPart.ArrayPoolBuffer;
+                
+                int totalRead = 0;
+                
+                // Read response stream into buffer based on ContentLength
+                while (totalRead < expectedBytes)
+                {
+                    // Calculate how many bytes we still need to read
+                    int remainingBytes = (int)(expectedBytes - totalRead);
+                    int bufferSpace = partBuffer.Length - totalRead;
```

Comment:
Is relying on the partBuffer's `Length` safe to do? The `byte[]` allocated for partBuffer via the ArrayPool Rent call which doesn't guarantee exact capacity size. It guarantees it will be at least as big as the request capacity. So if we get something larger then the capacity size which is essentially the part size won't relying on the rented byte [] throw off the counts.

Is the fix to use `downloadedPart.Length`?